packages/nimble-components/build/generate-workers/source/renderWorker.ts

```diff
@@ -0,0 +1,167 @@
+            value: new Float32Array()
+        });
+    private verticalScale: { a: number, b: number } = { a: 0, b: 1 };
+    private horizontalScale: { a: number, b: number } = { a: 0, b: 1 };
+    private margin: { top: number, right: number } = { top: 0, right: 0 };
+    private dieDimensions: { width: number, height: number } = { width: 1, height: 1 };
+    private colorCategories: { color: string, startValue: number, endValue?: number }[] = [];
+    private yLimits: { min: number, max: number } = { min: 0, max: 0 };
+    private xLimits: { min: number, max: number } = { min: 0, max: 0 };
+    private transform: { k: number, x: number, y: number } = { k: 1, x: 0, y: 0 };
+    private performanceTest: string | undefined;
+
+    public setCanvasDimensions(data: { width: number, height: number }): void {
+        this.canvas.width = data.width;
+        this.canvas.height = data.height;
+    }
+
+    public saveContext(): void {
+        this.context.save();
+    }
+
+    public restoreContext(): void {
+        this.context.restore();
+    }
+
+    public clearCanvas(): void {
+        this.context.clearRect(
+            0,
+            0,
+            this.canvas.width,
+            this.canvas.height
+        );
+    }
+
+    public scaleCanvas(data: { transform: { k: number, x: number, y: number } }): void {
+        this.context.translate(
+            data.transform.x,
+            data.transform.y
+        );
+        this.context.scale(
+            data.transform.k,
+            data.transform.k
+        );
+    }
+
+    public setCanvas(canvas: OffscreenCanvas): void {
+        this.canvas = canvas;
+        this.context = canvas.getContext('2d', {
+            willReadFrequently: true
+        })!;
+    }
+
+    public updateRenderConfig(
+        data: {
+            colorCategories: { color: string, startValue: number, endValue?: number }[],
+            verticalScale: { a: number, b: number },
+            horizontalScale: { a: number, b: number },
+            margin: { top: number, right: number },
+            xLimits: { min: number, max: number },
+            yLimits: { min: number, max: number },
+            transform: { k: number, x: number, y: number },
+            dieDimensions: { width: number, height: number }
+        }
+    ): void {
+        this.horizontalScale = data.horizontalScale;
+        this.verticalScale = data.verticalScale;
+        this.margin = data.margin;
+        this.yLimits = data.yLimits;
+        this.xLimits = data.xLimits;
+        this.dieDimensions = data.dieDimensions;
+        this.colorCategories = data.colorCategories;
+        this.transform = data.transform;
+    }
+
+    public emptyMatrix(): void {
+        this.dieTable = tableFromArrays({
+            colIndex: new Int32Array(),
+            rowIndex: new Int32Array(),
+            value: new Float32Array()
+        });
+    }
+
+    public rerenderDies(): void {
+        for (const tableRow of this.dieTable) {
```

Comment:
@rajsite the biggest time consumer seems to be [this iterator](https://github.com/apache/arrow/blob/a61f4af724cd06c3a9b4abd20491345997e532c0/js/src/table.ts#L220). There seems to be several parsing stages for this, so i think that may be an issue. I will look more into the visitor pattern and see if I can format the table to hold just the bare information for rendering and maybe  transform it into the underlying array buffers to access the values quicker

Reply:
Yea es6 iterators are pretty slow.. They don't offer another api to iterate quuicker?

Reply:
I tried different options but all of them were pretty slow, so eventually I got out of the table just the relevant columns and sent them over as array buffers. I also cached the column index positions so now the performance is comparable to before